### PR TITLE
feat(processors/k8sattributesprocessor): support metadata enrichment based on multiple attributes

### PR DESCRIPTION
**Pull request by sumo-drosiek**
**Description:**

This PR is continuation of #4310 as I cannot reopen it.

It deprecates current pod association structure and instead it proposes another one, which is able to handle multiple source attributes. The direct justification is to support metadata enrichment for pair (`k8s.pod.name`, `k8s.namespace.name`).

**Link to tracking Issue:** #4309

**Testing:** unit tests

**Documentation:** `doc.go`

### CODE REVIEW

File: processor/k8sattributesprocessor/doc.go [hunk 1]
```diff
@@ -53,6 +56,8 @@
 //   - k8s.node.name
 // Not all the attributes are guaranteed to be added. For example `k8s.cluster.name` usually is not provided by k8s API,
 // so likely it won't be set as an attribute.
+//
+// Only attibute names from `metadata` can be used for pod_association's `resource_attribute`
```
Comment (sumo-drosiek) [inline processor/k8sattributesprocessor/doc.go:60]:
```suggestion
// Only attribute names from `metadata` can be used for pod_association's `resource_attribute`
```

File: processor/k8sattributesprocessor/config.go [hunk 2]
```diff
@@ -73,7 +73,7 @@ type ExtractConfig struct {
 	// documentation for more details.
 	Annotations []FieldExtractConfig `mapstructure:"annotations"`
 
-	// Annotations allows extracting data from pod labels and record it
+	// Labels allows extracting data from pod labels and record it
```
Comment (sumo-drosiek) [inline processor/k8sattributesprocessor/config.go:76]:
```suggestion
	// Annotations allows extracting data from pod labels and record it
```

File: processor/k8sattributesprocessor/factory.go [hunk 3]
```diff
@@ -234,7 +235,7 @@ func warnDeprecatedMetadataConfig(logger *zap.Logger, cfg config.Processor) {
 			newName = conventions.AttributeK8SNodeName
 		}
 		if oldName != "" {
-			logger.Warn(fmt.Sprintf("%s has been deprecated in favor of %s for k8s-tagger processor", oldName, newName))
+			logger.Warn(fmt.Sprintf("%s has been deprecated in favor of %s for k8sattributes processor", oldName, newName))
```
Comment (sumo-drosiek) [inline processor/k8sattributesprocessor/factory.go:238]:
```suggestion
			logger.Warn(fmt.Sprintf("%s has been deprecated in favor of %s for k8s-tagger processor", oldName, newName))
```

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 4]
```diff
@@ -454,41 +457,90 @@ func (c *WatchClient) addOrUpdatePod(pod *api_v1.Pod) {
...
+				}
+
+				if attr != "" {
+					ret = append(ret, attr)
+				}
+			}
+		}
+
+		if len(ret) == len(assoc.Sources) {
+			ids = append(ids, PodIdentifier(strings.Join(ret, PodIdentifierDelimiter)))
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/internal/kube/client.go:498]: I don't think we can consider this as a reliable identifier given that now we allow using value of any attribute.

Let's say I have two sets of sources in `pod_association`, one uses values from `k8s.pod.name` attribute, another one uses `container.image`. It's theoretically possible that I have two different pods having the same value of `k8s.pod.name` attribute from one pod and `container.image` from another. 

I think we should use an array of attribute key-value pairs as an identifier.

File: processor/k8sattributesprocessor/factory.go [hunk 5]
```diff
@@ -250,3 +251,25 @@ func errWrongKeyConfig(cfg config.Processor) error {
...
+		if assoc.From != "" {
+			warnFrom = true
+		}
+		if assoc.Name != "" {
+			warnName = true
+		}
+	}
+
+	if warnFrom {
+		logger.Warn("pod_association.From has been deprecated in favor of pod_association.sources.From for k8sattributes processor")
```
Comment (pmm-sumo) [inline processor/k8sattributesprocessor/factory.go:269]: I think this warning is oriented around the code rather than the user-facing config and might be unclear to some. If we want o be nice, perhaps we could actually render the new config. I.e. look for all associations using the old format and they emit message akin to:

```
Deprecated pod_association configuration detected. Please replace:

pod_association:
  - ...
  
With:

pod_association:
  sources:
    - ...

File: processor/k8sattributesprocessor/internal/kube/client.go (hunk 4 above)
Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/internal/kube/client.go:498]: Yeah, that's true, but as we have to keep also `connection` type to be consistent I propose the following structure:

```
// PodAttributeIdentifier identifies pod attribute value
type PodAttributeIdentifier struct {
	key   AssociationSource
	value string
}

// PodIdentifier is a custom type to represent Pod by list of PodAttributeIdentifier
type PodIdentifier []PodAttributeIdentifier
```

Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/internal/kube/client.go:498]: I revisited my structure, and this will be better:
```
// PodIdentifier is a custom type to represent Pod by list of PodAttributeIdentifier
type PodIdentifier map[AssociationSource]string
```

Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/internal/kube/client.go:498]: but it won't work, as `incomparable map key type PodIdentifiercompiler[IncomparableMapKey](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal?utm_source%3Dgopls#IncomparableMapKey)`
:confused:

Comment (dmitryax) [reply processor/k8sattributesprocessor/internal/kube/client.go:498]: Array of structs should be a valid identifier. Something like this: `[maxSourcesLen]PodAssociationSourceConfig` where maxSourcesLen is max length across all the sources in `pod_association` list

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 6]
```diff
@@ -461,15 +461,25 @@ func (c *WatchClient) getPodFromAPI(pod *api_v1.Pod) *Pod {
...
 			case source.From == "connection":
-				if newPod.Address != "" {
-					ret = append(ret, newPod.Address)
+				if newPod.Address == "" {
+					skip = true
+				} else {
+					ret[i] = GetPodIdentifierAttribute(source, newPod.Address)
+					if source.Name != "" {
+						newPod.Attributes[source.Name] = newPod.Address
+					}
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/internal/kube/client.go:482]: Why is this needed? I think `source.Name` should be ignored if `source.From == "connection"`

File: processor/k8sattributesprocessor/internal/kube/kube.go [hunk 7]
```diff
@@ -34,11 +34,37 @@ const (
 	// MetadataFromPod is used to specify to extract metadata/labels/annotations from pod
 	MetadataFromPod = "pod"
 	// MetadataFromNamespace is used to specify to extract metadata/labels/annotations from namespace
-	MetadataFromNamespace = "namespace"
+	MetadataFromNamespace  = "namespace"
+	PodIdentifierMaxLength = 16
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/internal/kube/kube.go:38]: I think it can be even less to reduce memory footprint. Probably 4? It looks like the maximum number of resource attributes that we currently support https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/8465/commits/3f4811e94f1dc150edda717a81b01fe30f3ace61#diff-da945276e27b1e0a3eed6a7e4a97c5b1bb9d90da43bd8ef9fce8b533584d4292R486-R494

File: processor/k8sattributesprocessor/internal/kube/kube.go [hunk 8]
```diff
@@ -34,11 +34,37 @@ const (
...
+type PodIdentifierAttribute struct {
+	Source AssociationSource
+	Value  string
+}
+
+// PodIdentifier is a custom type to represent Pod identification
+type PodIdentifier [PodIdentifierMaxLength]PodIdentifierAttribute
+
+// GetPodIdentifierAttribute builds PodIdentifierAttribute using AssociationSource and value
+func GetPodIdentifierAttribute(source AssociationSource, value string) PodIdentifierAttribute {
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/internal/kube/kube.go:51]: Maybe `PodIdentifierAttributeFromSource`?

File: processor/k8sattributesprocessor/internal/kube/kube.go [hunk 9]
```diff
@@ -34,11 +34,37 @@ const (
...
+	return PodIdentifierAttribute{
+		Source: source,
+		Value:  value,
+	}
+}
+
+// GetPodIdentifierAttribute builds PodIdentifierAttribute using from, name (in order to build AssociationSource) and value
+func BuildPodIdentifierAttribute(from string, name string, value string) PodIdentifierAttribute {
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/internal/kube/kube.go:59]: Would be better to have `PodIdentifierAttributeFromConnection(value string)` and `PodIdentifierAttributeFromResourceAttribute(key string, value string)` instead?

File: processor/k8sattributesprocessor/internal/kube/client.go (hunk 4 above)
Comment (dmitryax) [reply processor/k8sattributesprocessor/internal/kube/client.go]: Sounds good. thanks

File: processor/k8sattributesprocessor/internal/kube/client.go (hunk 6 above)
Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/internal/kube/client.go:482]: Hmm, maybe it wasn't `source.Name` being used, just always `k8s.pod.ip` :sweat_smile:

File: processor/k8sattributesprocessor/factory.go (hunk 5 above)
Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/factory.go:269]:
```
2022-03-22T17:19:39.089+0100    warn    k8sattributesprocessor@v0.46.0/factory.go:272   Deprecated pod_association configuration detected. Please replace:
pod_association:
- from: resource_attribute
  name: k8s.pod.uid

with

pod_association:
- sources:
  - from: resource_attribute
  - name: k8s.pod.uid   {"kind": "processor", "name": "k8sattributes/2"}
```

File: processor/k8sattributesprocessor/internal/kube/kube.go (hunk 7 above)
Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/internal/kube/kube.go]: I agree, I would say that 2 is enough but set it to 4 just in case

File: processor/k8sattributesprocessor/internal/kube/kube.go (hunk 8 above)
Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/internal/kube/kube.go]: done

File: processor/k8sattributesprocessor/internal/kube/kube.go (hunk 9 above)
Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/internal/kube/kube.go]: done

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 10]
```diff
@@ -454,41 +457,90 @@ func (c *WatchClient) addOrUpdatePod(pod *api_v1.Pod) {
...
+				case conventions.AttributeK8SNamespaceName:
+					attr = newPod.Namespace
+				case conventions.AttributeK8SPodName:
+					attr = newPod.Name
+				case conventions.AttributeK8SPodUID:
+					attr = newPod.PodUID
+				case conventions.AttributeHostName:
+					attr = newPod.Address
+				default:
+					if v, ok := newPod.Attributes[source.Name]; ok {
```
[the remark this answers is not in the document]
Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/internal/kube/client.go]: We allow to use them in configuration, but using them won't work, as `newPod` doesn't have them

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3f4811e94f1dc150edda717a81b01fe30f3ace61/processor/k8sattributesprocessor/internal/kube/client.go#L275-L348

File: processor/k8sattributesprocessor/doc.go [hunk 11]
```diff
@@ -18,25 +18,28 @@
 // extracted metadata to the relevant spans, metrics and logs. The processor uses the kubernetes API to discover all pods
 // running in a cluster, keeps a record of their IP addresses, pod UIDs and interesting metadata.
 // The rules for associating the data passing through the processor (spans, metrics and logs) with specific Pod Metadata are configured via "pod_association" key.
-// It represents a list of rules that are executed in the specified order until the first one is able to do the match.
-// Each rule is specified as a pair of from (representing the rule type) and name (representing the extracted key name).
+// It represents a list of associations that are executed in the specified order until the first one is able to do the match.
+//
+// Each association is specified as a list of sources of association.
+// Sources represents list of rules. All rules are going to be executed and combination of result is going to be a pod metadata cache key.
```
[the remark this answers is not in the document]
Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/doc.go:24]: updated

**Comment by pmm-sumo** [comment]
I believe the items were addressed and this is ready for final review. @sumo-drosiek could you resolve the conflicts (there were some additional changes in the meantime)

@dmitryax can you have another look?

**Comment by newly12** [comment]
Hi all, how can we move forward on this PR?

File: processor/k8sattributesprocessor/pod_association.go [hunk 12]
```diff
@@ -32,44 +32,52 @@ import (
...
+					skip = true
 				}
+
+				ret[i] = kube.PodIdentifierAttributeFromSource(source, attributeValue)
 			}
 		}
+
+		// If all association sources has been resolved, return result
+		if !skip {
+			return asso.Name, ret
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/pod_association.go:67]: Isn't `asso.Name` deprecated? I'm not sure the first return argument even used anymore. Please update the function if it's not.

**Review by dmitryax — COMMENTED** [review]
Sorry for the delayed review

**Comment by sumo-drosiek** [comment]
@dmitryax Sorry for number of iterations. I cleaned up the code according to the latest comments

**Review by pmm-sumo — APPROVED** [review]

File: processor/k8sattributesprocessor/pod_association.go [hunk 13]
```diff
@@ -26,79 +26,91 @@ import (
...
-		// If association configured to take IP address from connection
-		switch {
-		case asso.From == "connection" && connectionIP != "":
-			return k8sIPLabelName, connectionIP
-		case asso.From == "resource_attribute":
-			// If association configured by resource_attribute
-			// In k8s environment, host.name label set to a pod IP address.
-			// If the value doesn't represent an IP address, we skip it.
-			if asso.Name == conventions.AttributeHostName {
-				if net.ParseIP(hostname) != nil {
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/pod_association.go:50]: Is this this part for parsing host.name of removed intentionally? This seems like unrelated change in the behavior. If it's done intentionally it probably should go through another PR with description why it should be removed

File: processor/k8sattributesprocessor/processor.go [hunk 14]
```diff
@@ -107,22 +107,21 @@ func (kp *kubernetesprocessor) processLogs(ctx context.Context, ld plog.Logs) (p
 
 // processResource adds Pod metadata tags to resource based on pod association configuration
 func (kp *kubernetesprocessor) processResource(ctx context.Context, resource pcommon.Resource) {
-	podIdentifierKey, podIdentifierValue := extractPodID(ctx, resource.Attributes(), kp.podAssociations)
-	if podIdentifierKey != "" {
-		resource.Attributes().InsertString(podIdentifierKey, string(podIdentifierValue))
+	podIdentifierValue := extractPodID(ctx, resource.Attributes(), kp.podAssociations)
+	for i := range podIdentifierValue {
+		if podIdentifierValue[i].Source.From == kube.ConnectionSource && podIdentifierValue[i].Value != "" {
+			resource.Attributes().InsertString(k8sIPLabelName, podIdentifierValue[i].Value)
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/processor.go:113]: probably we can add `break` here

File: processor/k8sattributesprocessor/processor.go [hunk 15]
```diff
@@ -107,22 +107,21 @@ func (kp *kubernetesprocessor) processLogs(ctx context.Context, ld plog.Logs) (p
...
 		return
 	}
 
-	if podIdentifierKey != "" {
-		if pod, ok := kp.kc.GetPod(podIdentifierValue); ok {
-			for key, val := range pod.Attributes {
-				resource.Attributes().InsertString(key, val)
-			}
-			kp.addContainerAttributes(resource.Attributes(), pod)
+	if pod, ok := kp.kc.GetPod(podIdentifierValue); ok {
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/processor.go:120]: I'm a bit concerned about removing the check for zero value here and using empty `PodIdentifier{}` that can be returned from `extractPodID`. Are we sure that `PodIdentifier{}` cannot get to the cache?

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 16]
```diff
@@ -454,41 +454,104 @@ func (c *WatchClient) addOrUpdatePod(pod *api_v1.Pod) {
...
 	}
 
+	return newPod
+}
+
+// getIdentifiersFromAssoc returns list of PodIdentifiers for given pod
+func (c *WatchClient) getIdentifiersFromAssoc(pod *Pod) []PodIdentifier {
+	if pod.Attributes == nil {
+		pod.Attributes = map[string]string{}
+	}
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/internal/kube/client.go:464]: I don't this part is needed. Let's avoid unnecessary allocations

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 17]
```diff
@@ -454,41 +454,104 @@ func (c *WatchClient) addOrUpdatePod(pod *api_v1.Pod) {
...
-	p, ok = c.GetPod(PodIdentifier(pod.UID))
-	c.m.RUnlock()
-
-	if ok && p.Name == pod.Name {
-		c.appendDeleteQueue(PodIdentifier(pod.UID), pod.Name)
+	newPod := c.getPodFromAPI(pod)
+	for _, id := range c.getIdentifiersFromAssoc(newPod) {
+		c.m.RLock()
+		p, ok := c.GetPod(id)
+		c.m.RUnlock()
```
Comment (newly12) [inline processor/k8sattributesprocessor/internal/kube/client.go:550]: Read Lock will be acquired in c.GetPod as well, should we remove Rlock & RUnlock here?

File: processor/k8sattributesprocessor/internal/kube/client.go [hunk 18]
```diff
@@ -454,41 +454,104 @@ func (c *WatchClient) addOrUpdatePod(pod *api_v1.Pod) {
...
+				case conventions.AttributeHostName:
+					attr = pod.Address
+				default:
+					if v, ok := pod.Attributes[source.Name]; ok {
+						attr = v
+					}
+				}
+
+				if attr == "" {
+					skip = true
```
Comment (newly12) [inline processor/k8sattributesprocessor/internal/kube/client.go:497]: probably we should add `break` here?

Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/internal/kube/client.go:426]: added

File: processor/k8sattributesprocessor/internal/kube/client.go (hunk 17 above)
Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/internal/kube/client.go]: yup, removed

File: processor/k8sattributesprocessor/internal/kube/client.go (hunk 16 above)
Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/internal/kube/client.go]: removed

File: processor/k8sattributesprocessor/pod_association.go (hunk 13 above)
Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/pod_association.go:50]: I reverted this change. Now it is more clear how it affects the flow.
The hostname is skipped if it is not an ip address. It is not obvious for me, but cleaning it up can be part of another PR, as you said

File: processor/k8sattributesprocessor/processor.go (hunk 14 above)
Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/processor.go:113]: added

File: processor/k8sattributesprocessor/processor.go (hunk 15 above)
Comment (sumo-drosiek) [reply processor/k8sattributesprocessor/processor.go]: I reverted this change. TBH, same behavior like empty `PodIdentifier{}` is going to be possible with configuring only one attribute e.g. `namespace` as association source

**Comment by newly12** [comment]
Hi, could this PR be reviewed again? thanks!

**Comment by Wolodija** [comment]
```
 FAIL
FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/splunkhecexporter	35.898s
FAIL
```
This fail is unrelated. I rebased branch on the top of the main

**Comment by sumo-drosiek** [comment]
@dmitryax Could you take a look at it?

**Comment by newly12** [comment]
@dmitryax Could you take a look at it?

**Comment by dmitryax** [comment]
Hi. Sorry for the delay. I will review it on Monday

**Review by dmitryax — COMMENTED** [review]
Found just couple nits after a quick review. I will do another thorough review tomorrow since it's a pretty significant change

**Comment by dmitryax** [comment]
Just some nits. Otherwise LGTM. Please rebase

**Review by dmitryax — APPROVED** [review]

File: unreleased/drosiek-k8sprocessor.yaml [hunk 19]
```diff
@@ -0,0 +1,11 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: deprecation
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: k8sattributesprocessor
+
+# A brief description of the change
+note: Announcing `pod_association` rules will be deprecated. Use `pod_associations.sources` rules instead. This is in order to support metadata enrichment based on multiple attributes.
```
Comment (dmitryax) [inline unreleased/drosiek-k8sprocessor.yaml:8]:
```suggestion
note: Announcing `pod_association` rules will be deprecated. Use `pod_association.sources` rules instead. This is in order to support metadata enrichment based on multiple attributes.
```
